platforms/hip/src/HipContext.cpp

```diff
@@ -0,0 +1,942 @@
+        useDoublePrecision = false;
+        useMixedPrecision = true;
+    }
+    else if (precision == "double") {
+        useDoublePrecision = true;
+        useMixedPrecision = false;
+    }
+    else
+        throw OpenMMException("Illegal value for Precision: "+precision);
+    char* cacheVariable = getenv("OPENMM_CACHE_DIR");
+    cacheDir = (cacheVariable == NULL ? tempDir : string(cacheVariable));
+#ifdef WIN32
+    this->tempDir = tempDir+"\\";
+    cacheDir = cacheDir+"\\";
+#else
+    this->tempDir = tempDir+"/";
+    cacheDir = cacheDir+"/";
+#endif
+    contextIndex = platformData.contexts.size();
+    string errorMessage = "Error initializing Context";
+    if (originalContext == NULL) {
+        isLinkedContext = false;
+        int numDevices;
+        CHECK_RESULT(hipGetDeviceCount(&numDevices));
+        if (deviceIndex < -1 || deviceIndex >= numDevices)
+            throw OpenMMException("Illegal value for DeviceIndex: "+intToString(deviceIndex));
+
+        vector<int> devicePrecedence;
+        if (deviceIndex == -1) {
+            devicePrecedence = getDevicePrecedence();
+        } else {
+            devicePrecedence.push_back(deviceIndex);
+        }
+
+        this->deviceIndex = -1;
+        for (int i = 0; i < static_cast<int>(devicePrecedence.size()); i++) {
+            int trialDeviceIndex = devicePrecedence[i];
+            CHECK_RESULT(hipDeviceGet(&device, trialDeviceIndex));
+            // try setting device
+            if (hipSetDevice(device) == hipSuccess) {
+                this->deviceIndex = trialDeviceIndex;
+                break;
+            }
+
+        }
+        if (this->deviceIndex == -1) {
+            if (deviceIndex != -1)
+                throw OpenMMException("The requested HIP device could not be loaded");
+            else
+                throw OpenMMException("No compatible HIP device is available");
+        }
+        CHECK_RESULT(hipStreamCreateWithFlags(&defaultStream, hipStreamNonBlocking));
+    }
+    else {
+        isLinkedContext = true;
+        this->deviceIndex = originalContext->deviceIndex;
+        this->device = originalContext->device;
+        defaultStream = originalContext->defaultStream;
+    }
+    currentStream = defaultStream;
+
+    hipDeviceProp_t props;
+    CHECK_RESULT(hipGetDeviceProperties(&props, device));
+
+    // set device properties
+    this->simdWidth = props.warpSize;
+    this->sharedMemPerBlock = props.sharedMemPerBlock;
+
+    gpuArchitecture = props.gcnArchName;
+    // HIP-TODO: find a good value here
+    int numThreadBlocksPerComputeUnit = 6;
+
+    if (gpuArchitecture.find("gfx908") == 0 ||
+        gpuArchitecture.find("gfx90a") == 0 ||
+        gpuArchitecture.find("gfx94") == 0) {
+        // MI100 and newer CDNA support 32 bit float atomic add
+        // HIP-TODO: evaluate performance on RDNA 3 (gfx11* and gfx12*), where it's supported too
+        this->supportsHardwareFloatGlobalAtomicAdd = true;
+    }
```

Comment:
Will this need to be regularly updated as new GPUs get released?  I wonder if there's a more future-proof way of determining whether it's supported?

Reply:
Changed